File: src/pizza_types.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "../include/pizza_types.h"

int load_pizza_menu(PizzaMenu *menu, const char *filename) {
    FILE *file = fopen(filename, "r");
    if (!file) {
        // If file doesn't exist, initialize with default menu
        init_default_menu(menu);
        return 0;
    }
    
    menu->count = 0;
    char line[100];
    
    while (fgets(line, sizeof(line), file) && menu->count < MAX_PIZZA_TYPES) {
        char name[MAX_PIZZA_NAME_LENGTH];
        double price;
        int available;
        
        if (sscanf(line, "%49[^,],%lf,%d", name, &price, &available) == 3) {
            strcpy(menu->pizzas[menu->count].name, name);
            menu->pizzas[menu->count].price = price;
            menu->pizzas[menu->count].available = available;
            menu->count++;
        }
    }
    
    fclose(file);
    return 1;
}

int save_pizza_menu(const PizzaMenu *menu, const char *filename) {
    FILE *file = fopen(filename, "w");
    if (!file) {
        return 0;
    }
    
    for (int i = 0; i < menu->count; i++) {
        fprintf(file, "%s,%.2f,%d\n", 
                menu->pizzas[i].name, 
                menu->pizzas[i].price, 
                menu->pizzas[i].available);
    }
    
    fclose(file);
    return 1;
}

void add_pizza_type(PizzaMenu *menu, const char *name, double price) {
    if (menu->count >= MAX_PIZZA_TYPES) {
        printf("Error: Maximum number of pizza types reached.\n");
        return;
    }
    
    // Check if pizza already exists
    if (find_pizza_by_name(menu, name) != -1) {
        printf("Error: Pizza type '%s' already exists.\n", name);
        return;
    }
    
    strcpy(menu->pizzas[menu->count].name, name);
    menu->pizzas[menu->count].price = price;
    menu->pizzas[menu->count].available = 1;
    menu->count++;
    
    printf("Pizza type '%s' added successfully!\n", name);
}

void display_pizza_menu(const PizzaMenu *menu) {
    printf("\n=== Pizza Menu ===\n");
    printf("%-3s %-25s %-10s %-10s\n", "ID", "Pizza Name", "Price", "Available");
    printf("--------------------------------------------------\n");
    
    for (int i = 0; i < menu->count; i++) {
        printf("%-3d %-25s €%-9.2f %-10s\n", 
               i + 1,
               menu->pizzas[i].name, 
               menu->pizzas[i].price,
               menu->pizzas[i].available ? "Yes" : "No");
    }
    printf("\n");
}

int find_pizza_by_name(const PizzaMenu *menu, const char *name) {
    for (int i = 0; i < menu->count; i++) {
        if (strcasecmp(menu->pizzas[i].name, name) == 0) {
            return i;
        }
    }
    return -1;
}

void remove_pizza_type(PizzaMenu *menu, const char *name) {
    int index = find_pizza_by_name(menu, name);
    if (index == -1) {
        printf("Error: Pizza type '%s' not found.\n", name);
        return;
    }
    
    // Shift all elements after the removed one
    for (int i = index; i < menu->count - 1; i++) {
        menu->pizzas[i] = menu->pizzas[i + 1];
    }
    
    menu->count--;
    printf("Pizza type '%s' removed successfully!\n", name);
}

void init_default_menu(PizzaMenu *menu) {
    menu->count = 0;
    
    add_pizza_type(menu, "Margherita", 10.00);
    add_pizza_type(menu, "Pepperoni", 12.50);
    add_pizza_type(menu, "Hawaiian", 13.00);
    add_pizza_type(menu, "Quattro Stagioni", 14.50);
    add_pizza_type(menu, "Vegetarian", 11.50);
    add_pizza_type(menu, "Meat Lovers", 16.00);
}

File: python-demo/src/pizza_types.py
"""
Pizza Types Module
Handles pizza menu management with secure data validation
"""

import os
import json
import re
from typing import List, Dict, Optional, Tuple
from decimal import Decimal, InvalidOperation

# Security constants
MAX_PIZZA_NAME_LENGTH = 50
MAX_PIZZA_TYPES = 20
MAX_PRICE = Decimal('999.99')
MIN_PRICE = Decimal('0.01')
MAX_QUANTITY = 1000

class PizzaType:
    """Represents a single pizza type with validation"""
    
    def __init__(self, name: str, price: Decimal, available: bool = True):
        self.name = self._validate_name(name)
        self.price = self._validate_price(price)
        self.available = bool(available)
    
    def _validate_name(self, name: str) -> str:
        """Validate and sanitize pizza name"""
        if not isinstance(name, str):
            raise ValueError("Pizza name must be a string")
        
        # Strip whitespace and validate length
        name = name.strip()
        if not name:
            raise ValueError("Pizza name cannot be empty")
        
        if len(name) > MAX_PIZZA_NAME_LENGTH:
            raise ValueError(f"Pizza name too long (max {MAX_PIZZA_NAME_LENGTH} chars)")
        
        # Sanitize - allow only alphanumeric, spaces, and basic punctuation
        if not re.match(r'^[a-zA-Z0-9\s\-\'\&\.]+$', name):
            raise ValueError("Pizza name contains invalid characters")
        
        return name
    
    def _validate_price(self, price) -> Decimal:
        """Validate and convert price to Decimal"""
        try:
            price_decimal = Decimal(str(price))
        except (InvalidOperation, ValueError, TypeError):
            raise ValueError("Invalid price format")
        
        if price_decimal < MIN_PRICE or price_decimal > MAX_PRICE:
            raise ValueError(f"Price must be between €{MIN_PRICE} and €{MAX_PRICE}")
        
        return price_decimal.quantize(Decimal('0.01'))
    
    def to_dict(self) -> Dict:
        """Convert to dictionary for serialization"""
        return {
            'name': self.name,
            'price': str(self.price),
            'available': self.available
        }
    
    @classmethod
    def from_dict(cls, data: Dict) -> 'PizzaType':
        """Create PizzaType from dictionary"""
        return cls(
            name=data['name'],
            price=Decimal(data['price']),
            available=data.get('available', True)
        )

class PizzaMenu:
    """Manages the pizza menu with secure operations"""
    
    def __init__(self):
        self.pizzas: List[PizzaType] = []
        self._init_default_menu()
    
    def _init_default_menu(self):
        """Initialize with default pizza types"""
        default_pizzas = [
            ('Margherita', '10.00'),
            ('Pepperoni', '12.50'),
            ('Hawaiian', '13.00'),
            ('Quattro Stagioni', '14.50'),
            ('Vegetarian', '11.50'),
            ('Meat Lovers', '16.00')
        ]
        
        for name, price in default_pizzas:
            try:
                self.add_pizza_type(name, Decimal(price))
            except ValueError as e:
                print(f"Warning: Could not add default pizza {name}: {e}")
    
    def add_pizza_type(self, name: str, price) -> bool:
        """Add a new pizza type with validation"""
        if len(self.pizzas) >= MAX_PIZZA_TYPES:
            raise ValueError(f"Maximum number of pizza types ({MAX_PIZZA_TYPES}) reached")
        
        # Check if pizza already exists
        if self.find_pizza_by_name(name) is not None:
            raise ValueError(f"Pizza type '{name}' already exists")
        
        pizza = PizzaType(name, price)
        self.pizzas.append(pizza)
        return True
    
    def remove_pizza_type(self, name: str) -> bool:
        """Remove a pizza type by name"""
        pizza = self.find_pizza_by_name(name)
        if pizza is None:
            raise ValueError(f"Pizza type '{name}' not found")
        
        self.pizzas.remove(pizza)
        return True
    
    def find_pizza_by_name(self, name: str) -> Optional[PizzaType]:
        """Find pizza by name (case-insensitive)"""
        name_lower = name.lower().strip()
        for pizza in self.pizzas:
            if pizza.name.lower() == name_lower:
                return pizza
        return None
    
    def get_pizza_by_index(self, index: int) -> Optional[PizzaType]:
        """Get pizza by index (1-based)"""
        if 1 <= index <= len(self.pizzas):
            return self.pizzas[index - 1]
        return None
    
    def display_menu(self):
        """Display formatted pizza menu"""
        print("\n=== Pizza Menu ===")
        print(f"{'ID':<3} {'Pizza Name':<25} {'Price':<10} {'Available':<10}")
        print("-" * 50)
        
        for i, pizza in enumerate(self.pizzas, 1):
            availability = "Yes" if pizza.available else "No"
            print(f"{i:<3} {pizza.name:<25} €{pizza.price:<9} {availability:<10}")
        print()
    
    def load_from_file(self, filename: str) -> bool:
        """Load menu from JSON file with security checks"""
        try:
            # Validate file path
            if not self._is_safe_path(filename):
                raise ValueError("Invalid file path")
            
            if not os.path.exists(filename):
                return False
            
            # Check file size (prevent DoS)
            if os.path.getsize(filename) > 1024 * 1024:  # 1MB limit
                raise ValueError("File too large")
            
            with open(filename, 'r', encoding='utf-8') as f:
                data = json.load(f)
            
            if not isinstance(data, list):
                raise ValueError("Invalid file format")
            
            # Validate and load pizzas
            new_pizzas = []
            for item in data:
                if not isinstance(item, dict):
                    continue
                try:
                    pizza = PizzaType.from_dict(item)
                    new_pizzas.append(pizza)
                except ValueError:
                    continue  # Skip invalid entries
            
            if new_pizzas:
                self.pizzas = new_pizzas
            
            return True
            
        except (json.JSONDecodeError, OSError, ValueError) as e:
            print(f"Error loading menu: {e}")
            return False
    
    def save_to_file(self, filename: str) -> bool:
        """Save menu to JSON file with security checks"""
        try:
            # Validate file path
            if not self._is_safe_path(filename):
                raise ValueError("Invalid file path")
            
            # Ensure directory exists
            os.makedirs(os.path.dirname(filename), mode=0o755, exist_ok=True)
            
            data = [pizza.to_dict() for pizza in self.pizzas]
            
            # Write to temporary file first (atomic operation)
            temp_filename = filename + '.tmp'
            with open(temp_filename, 'w', encoding='utf-8') as f:
                json.dump(data, f, indent=2, ensure_ascii=False)
            
            # Move to final location
            os.rename(temp_filename, filename)
            return True
            
        except (OSError, ValueError) as e:
            print(f"Error saving menu: {e}")
            return False
    
    def _is_safe_path(self, path: str) -> bool:
        """Validate file path for security"""
        # Normalize path
        norm_path = os.path.normpath(path)
        
        # Check for path traversal attempts
        if '..' in norm_path or norm_path.startswith('/'):
            return False
        
        # Only allow files in data directory
        if not norm_path.startswith('data/'):
            return False
        
        return True
    
    def count(self) -> int:
        """Get number of pizza types"""
        return len(self.pizzas)
